engine/table/src/main/java/io/deephaven/engine/table/impl/multijoin/StaticMultiJoinStateManagerTypedBase.java

```diff
@@ -0,0 +1,218 @@
+import io.deephaven.engine.table.impl.util.LongColumnSourceWritableRowRedirection;
+import io.deephaven.engine.table.impl.util.TypedHasherUtil;
+import io.deephaven.engine.table.impl.util.TypedHasherUtil.BuildOrProbeContext.BuildContext;
+import io.deephaven.engine.table.impl.util.WritableRowRedirection;
+import io.deephaven.util.QueryConstants;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static io.deephaven.engine.table.impl.JoinControl.CHUNK_SIZE;
+import static io.deephaven.engine.table.impl.JoinControl.MAX_TABLE_SIZE;
+import static io.deephaven.engine.table.impl.util.TypedHasherUtil.getKeyChunks;
+
+public abstract class StaticMultiJoinStateManagerTypedBase implements MultiJoinStateManager {
+    protected final ColumnSource<?>[] keySourcesForErrorMessages;
+    private final List<LongArraySource> indexSources = new ArrayList<>();
+
+    public static final long NO_RIGHT_STATE_VALUE = RowSet.NULL_ROW_KEY;
+    public static final long EMPTY_RIGHT_STATE = QueryConstants.NULL_LONG;
+    public static final long DUPLICATE_RIGHT_STATE = -2;
+
+    // The number of slots in our hash table.
+    protected int tableSize;
+
+    // The number of entries in our hash table in use.
+    protected long numEntries = 0;
+
+    // The table will be rehashed to a load factor of targetLoadFactor if our loadFactor exceeds maximumLoadFactor
+    // or if it falls below minimum load factor we will instead contract the table.
+    private final double maximumLoadFactor;
+
+    // The keys for our hash entries.
+    protected final ChunkType[] chunkTypes;
+    protected final WritableColumnSource[] mainKeySources;
+
+    // The output sources representing the keys of our joined table.
+    protected final WritableColumnSource[] outputKeySources;
+
+    // Store sentinel information and maps hash slots to output row keys.
+    protected ImmutableLongArraySource slotToOutputRow = new ImmutableLongArraySource();
+
+    protected StaticMultiJoinStateManagerTypedBase(ColumnSource<?>[] tableKeySources,
+            ColumnSource<?>[] keySourcesForErrorMessages,
+            int tableSize,
+            double maximumLoadFactor) {
+        this.keySourcesForErrorMessages = keySourcesForErrorMessages;
+
+        this.tableSize = tableSize;
+        Require.leq(tableSize, "tableSize", MAX_TABLE_SIZE);
+        Require.gtZero(tableSize, "tableSize");
+        Require.eq(Integer.bitCount(tableSize), "Integer.bitCount(tableSize)", 1);
+        Require.inRange(maximumLoadFactor, 0.0, 0.95, "maximumLoadFactor");
+
+        mainKeySources = new WritableColumnSource[tableKeySources.length];
+        chunkTypes = new ChunkType[tableKeySources.length];
+
+        outputKeySources = new WritableColumnSource[tableKeySources.length];
+
+        for (int ii = 0; ii < tableKeySources.length; ++ii) {
+            chunkTypes[ii] = tableKeySources[ii].getChunkType();
+            mainKeySources[ii] = InMemoryColumnSource.getImmutableMemoryColumnSource(tableSize,
```

Comment:
These aren't immutable.

Reply:
The pattern for OA is to have these Immutable, then adjust size and contents during rehash.